### PR TITLE
RandWord: Add definition for random word 

### LINKED ISSUE #2807: Random Words: show definition of result
Could we also show [definition](https://wordnik.com/words/brangle) of (random word) result to the user? This might help the end-user to use `Random Words` IA for vocabulary improvement.

------
IA Page:  http://duck.co/ia/view/rand_word 
[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @hunterlang

### PR DESCRIPTION
**Pull request by VeerpalBrar**
###### What does your Pull Request do (check all that apply)?

Choose the most relevant items and use the following title template to name
your Pull Request.

- [ ] New Instant Answer: **`New {IA Name} Instant Answer`**
- [x] Improvement
    - [ ] Bug fix: **`{IA Name}: Fix {Issue number or one-line description}`**
    - [x] Enhancement: **`{IA Name}: {Description of Improvements}`**
- [ ] Non–Instant Answer
    - [ ] Other (Role, Template, Test, Documentation, etc.): **`{SpiceRole/Templates/Tests/Docs}: {Short Description}`**

###### Description of changes

When users searches for an random word, the definition is now provided along with the word. 

###### Which issues (if any) does this fix?

Fixes #2807 by adding the definition functionality. 

###### People to notify (@mention interested parties)

---

Instant Answer Page: http://duck.co/ia/view/rand_word 

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @hunterlang

### CODE REVIEW

**Comment by daxtheduck** [comment]
@VeerpalBrar Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by gaulrobe** [comment]
@VeerpalBrar just curious, why the new `data` directory here?

**Comment by VeerpalBrar** [comment]
@gaulrobe The enhancement requires the use of another API endpoint. Each endpoint needs it own callback function. I needed to create the data folder so that one endpoint would call back to function ddg_spice_rand_word_data and the other to function ddg_spice_rand_word_fetch_to. 

I think data might not have been the best name, since the folder isn't exactly data related.

File: share/spice/rand_word/data/rand_word_data.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+(function(env) {
+    "use strict";
+    env.ddg_spice_rand_word_data = function(api_result) {
+        if(!api_result || api_result.length === 0) {
+            return Spice.failed('rand_word');
+        }
+        console.log(api_result)
```
Comment (gaulrobe) [inline share/spice/rand_word/data/rand_word_data.js]: This console log can probably go away.

File: share/spice/rand_word/data/rand_word_data.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+            },
+            templates: {
+                group: 'list',
+                options: {
+                    list_content: Spice.rand_word_data.content,
+                    moreAt: true
+                }
+            }
+        };
+        if(list_of_list[0].length == 1) {
```
Comment (gaulrobe) [inline share/spice/rand_word/data/rand_word_data.js]: This if-block has some uneven tabbing, could you fix it up for better readability?

File: share/spice/rand_word/data/rand_word_data.js (hunk 1 above)
Comment (VeerpalBrar) [inline share/spice/rand_word/data/rand_word_data.js]: opps! That was there for testing purposes.

File: share/spice/rand_word/data/single_word.handlebars [hunk 3]
```diff
@@ -0,0 +1,2 @@
+<a href="http://wordnik.com/words/{{word}}">{{word}}</a>
```
Comment (gaulrobe) [inline share/spice/rand_word/data/single_word.handlebars]: Do we need the handlebars template anymore? Can we leverage a `title` and `subtitle` setting in the javascript here?

Similar to https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/share/spice/xkcd/display/xkcd_display.js

**Comment by moollaza** [comment]
> I needed to create the data folder so that one endpoint would call back to function ddg_spice_rand_word_data and the other to function ddg_spice_rand_word_fetch_to.

@VeerpalBrar creating a subdirectory isn't exactly necessary. You can just namespace the `alt_to` endpoint like so: `rand_word_fetch_id`

Then you'd have `return $.getJSON("/js/spice/rand_word_fetch_id/" + word);`

That approach should really simplify things

**Comment by moollaza** [comment]
@VeerpalBrar actually -- you don't even need to make a new `alt_to` -- the Definition spice already has it's own endpoint which you can call from the JS directly!

**Comment by VeerpalBrar** [comment]
@moollaza I deciding to just change the namespace of the alt_to. When I tried to definition spice endpoint it said that it could not be found.

**Comment by moollaza** [comment]
> When I tried to definition spice endpoint it said that it could not be found.

@VeerpalBrar Yeah, I just realized that `/js/spice/dictionary/definition/` won't work work for us because it's going to wrap the JSON response in a function call for the Dictionary Spice because that Spice defines a `callback` param. 

`alt_to` is the way to go here 👍

File: share/spice/rand_word/rand_word.js [hunk 4]
```diff
@@ -34,23 +31,38 @@
...
-                word: first.word
-            };
-            
-            spiceObj.templates = {
-                group: 'text',
-                options: {
-                    content: Spice.rand_word.single_word,
-                    moreAt: true
+        if (list_of_list[0].length == 1) {
+            get_definition(first.word).done(function(api_result) {
```
Comment (moollaza) [inline share/spice/rand_word/rand_word.js]: the `get_definition` function doesn't seem to serve much purpose. Let's just use the canonical form:

```js
$.getJSON("/js/spice/rand_word_fetch_id/" + word, function(data) {
    var first = data[0];
    $.extend(spiceObj, {
        data: { .... },
        templates: { .... },
        normalize: funciton() { ... }
    });

    Spice.add(spiceObj);
}
```

**Comment by VeerpalBrar** [comment]
If there are no more suggestions, can this be merged now?

File: lib/DDG/Spice/RandWord.pm [hunk 5]
```diff
@@ -11,6 +11,12 @@ spice proxy_cache_valid => "418 1d";
 
 triggers any => "random word", "random words";
 
+spice alt_to => {
+	rand_word_fetch_id => {
+		to => 'http://api.wordnik.com:80/v4/word.json/$1/definitions?limit=1&includeRelated=false&sourceDictionaries=all&useCanonical=false&includeTags=false&api_key={{ENV{DDG_SPICE_WORDNIK_APIKEY}}}',
```
Comment (GuiltyDolphin) [inline lib/DDG/Spice/RandWord.pm]: Lets convert these tabs to spaces (4 spaces per indent).

File: share/spice/rand_word/rand_word.js [hunk 6]
```diff
@@ -34,23 +31,38 @@
...
+
+                $.extend(spiceObj, {
+
+                    data: {
+                        title: "Random Word",
+                        word: first.word,
+                        definition: first.text
+                    },
+                    templates: {
+                        group: 'text',
```
Comment (GuiltyDolphin) [inline share/spice/rand_word/rand_word.js]: Trailing comma.

File: share/spice/rand_word/rand_word.js [hunk 7]
```diff
@@ -34,23 +31,38 @@
...
+
```
Comment (GuiltyDolphin) [inline share/spice/rand_word/rand_word.js]: Lets lose this line.

File: share/spice/rand_word/rand_word.js [hunk 8]
```diff
@@ -34,23 +31,38 @@
...
+            });
+        } else {
+            Spice.add(spiceObj);
         }
-        
-        Spice.add( spiceObj );
-        
     }
-}(this));
+}(this));
```
Comment (GuiltyDolphin) [inline share/spice/rand_word/rand_word.js]: Add newline at end of the file.

File: share/spice/rand_word/rand_word.js [hunk 9]
```diff
@@ -34,23 +31,34 @@
...
-                    content: Spice.rand_word.single_word,
-                    moreAt: true
-                }
-            };
+        if (list_of_list[0].length == 1) {
+            $.getJSON("/js/spice/rand_word_fetch_id/" + first.word, function(data) {
+                var first = data[0];
+                $.extend(spiceObj, {
+                    data: {
+                        title: "Random Word",
```
Comment (moollaza) [inline share/spice/rand_word/rand_word.js]: @VeerpalBrar let's make the title the actual random word

We can use a subtitle to specify it was a randomly chosen word

Comment (moollaza) [inline share/spice/rand_word/rand_word.js]: You can make the `title` into a clickable link if you define a `url` property 👍

**Comment by VeerpalBrar** [comment]
@moollaza I decided to leave out putting random word in the subtitle because if the user searches for 'random word' is it self explanatory that the word generated is random.  

Also, I could not figure out how to make the title clickable.

**Comment by moollaza** [comment]
@VeerpalBrar I've deployed the updates to Beta. I think it's OK, but I find myself wanting to click in title to see more info.

It also seems like we get a _lot_ of words that are a plural form of another word. Unfortunately the definition is always "Plural form of $word", which leaves me wanting to click the singular word.

Is there anyway we can avoid pluralized results so we always get a word with a proper definition?

**Comment by moollaza** [comment]
Merging this for now (been ongoing for too long!), but there's still more room for improvement 👍

**Pull request merged**